pkg/policy/l4.go

```diff
@@ -1039,7 +1126,29 @@ func createL4Filter(policyCtx PolicyContext, peerEndpoints api.EndpointSelectorS
 	}
 
 	if l7Parser != ParserTypeNone || auth != nil || policyCtx.IsDeny() {
-		l4.PerSelectorPolicies.addPolicyForSelector(l7Parser, rules, terminatingTLS, originatingTLS, auth, policyCtx.IsDeny(), sni, listener, priority)
+		modifiedRules := rules
+
+		// If we have L7 rules and default deny is disabled (EnableDefaultDeny=false), we should ensure those rules
+		// don't cause other L7 traffic to be denied.
+		// Special handling for L7 rules is applied when:
+		// 1. We have L7 rules
+		// 2. Default deny is disabled for this direction
+		// 3. This is a positive policy (not a deny policy)
+		hasL7Rules := !rules.IsEmpty()
+		isDefaultDenyDisabled := (ingress && !policyCtx.DefaultDenyIngress()) || (!ingress && !policyCtx.DefaultDenyEgress())
+		isAllowPolicy := !policyCtx.IsDeny()
+
+		if hasL7Rules && isDefaultDenyDisabled && isAllowPolicy {
+			logger.Debug("Adding wildcard L7 rules for default-allow policy",
+				logfields.L7Parser, l7Parser,
+				logfields.Ingress, ingress,
+				logfields.DefaultDenyIngress, policyCtx.DefaultDenyIngress(),
```

Comment:
Could probably drop the DefaultDenyIngress / Egress logfields, if you like; they're clear from the condition above.